### PR TITLE
HOTFIX: Cherrypicking state deadlock fix from 0.11

### PR DESCRIPTION
**Pull request by enothereska**
Cherrypicking additional changes made as part of this PR https://github.com/apache/kafka/pull/3622, back to trunk.

### CODE REVIEW

**Comment by enothereska** [comment]
Thanks @dguy this should complete all state transitions (finally)

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6871/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6857/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by enothereska** [comment]
System test passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1043/

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -679,13 +690,7 @@ public void close() {
...
-            if (state.isCreatedOrRunning() || state == ERROR) {
-                state = PENDING_SHUTDOWN;
-                return true;
-            }
-            return false;
-        }
+        return setState(PENDING_SHUTDOWN);
     }
 
     private void closeGlobalStreamThread() {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:696]: Do we need to do `globalStreamThread.setStateListener(null)` here?

**Review by dguy — COMMENTED** [review]
Thanks @enothereska, just one comment about the `GlobalStreamThread`

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 1 above)
Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:696]: Yes, good catch, thanks!

Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:696]: Done.

**Review by dguy — APPROVED** [review]
LGTM

**Comment by dguy** [comment]
merged to trunk

**Pull request closed without merging**